README.md
# date_calc

Date manipulation library for easy calculating.

## Usage

### Installation

Add `date_calc` as a dependency in pubspec.yaml file.

```yaml
dependencies:
  date_calc: ^0.1.0
```

Run `pub install`.

### Using

```dart
import 'package:date_calc/date_calc.dart';

final date = DateCalc(2019, 1, 5);
print(date.addMonth(1));
// => 2019-02-05 00:00:00.000
print(date.addYear(1).isLeapYear());
// => true
print(date.addYear(1).addMonth(1).daysInMonth());
// => 29
print(date.dup(minute: 11));
// => 2019-01-05 00:11:00.000
print(date.differenceValue(date: DateTime(2019, 3, 3), type: DateType.month));
// => 1
print(DateCalc.now().isToday());
// => true
```

Because DateCalc is a subclass of DateTime, you can use all DateTime methods.  
When you need exactly DateTime object, use `toDate()` method.

```dart
final date = DateCalc(2019, 1, 5);
date.subtractYear(1).toDate();
```

## milestones for 1.0.0

- support for DST
- support for UTC & Timezone
- support for leapsecond


CHANGELOG.md
## 0.2.0

* Null-Safety for Dart ^2.12.0

## 0.1.1+4

* Updated intl-dependency (to use with package datetime_picker_formfield 2.0.0)
* Updated test-dependency

## 0.1.1+3

* Fix to `toDate` method returns correct `isUtc` parameter.

## 0.1.1+2

* Add remaining add & subtract methods.

## 0.1.1+1

* Remove [initializeDateFormatting] from toFormattedString.  
As it affects for app, it should be called from app codes.

## 0.1.1

* Add parse method.
* Add toFormattedString method.

## 0.1.0+1

* Fix dependency version.

## 0.1.0

* Initial Open Source release.

test/date_calc_test.dart
import 'package:date_calc/date_calc.dart';
import 'package:intl/date_symbol_data_local.dart';
import 'package:test/test.dart';

void main() {
  test('Yesterday isSameDay yesterday, yesrterday isNotSameDay today', () {
    final t = DateTime.now();
    expect(
        DateCalc(t.year, t.month, t.day - 1)
            .isSameDay(DateTime(t.year, t.month, t.day - 1)),
        true);
    expect(DateCalc(t.year, t.month, t.day - 1).isSameDay(t), false);
    expect(
        DateCalc(t.year, t.month, t.day - 1)
            .isNotSameDay(DateTime(t.year, t.month, t.day - 1)),
        false);
    expect(DateCalc(t.year, t.month, t.day - 1).isNotSameDay(t), true);
  });

  test('Yesterday !isToday, today isToday', () {
    final t = DateTime.now();
    expect(DateCalc(t.year, t.month, t.day - 1).isToday(), false);
    expect(DateCalc(t.year, t.month, t.day).isToday(), true);
  });

  test('isLeapYear goes well', () {
    expect(DateCalc(2019).isLeapYear(), false);
    expect(DateCalc(2020).isLeapYear(), true);
    expect(DateCalc(2100).isLeapYear(), false);
    expect(DateCalc(2000).isLeapYear(), true);
    expect(DateCalc.isLeapYearFor(2000), true);
  });

  test('daysInMonth returns days in month', () {
    expect(DateCalc(2019, 1).daysInMonth(), 31);
    expect(DateCalc(2019, 2).daysInMonth(), 28);
    expect(DateCalc(2019, 3).daysInMonth(), 31);
    expect(DateCalc(2019, 4).daysInMonth(), 30);
    expect(DateCalc(2019, 5).daysInMonth(), 31);
    expect(DateCalc(2019, 6).daysInMonth(), 30);
    expect(DateCalc(2019, 7).daysInMonth(), 31);
    expect(DateCalc(2019, 8).daysInMonth(), 31);
    expect(DateCalc(2019, 9).daysInMonth(), 30);
    expect(DateCalc(2019, 10).daysInMonth(), 31);
    expect(DateCalc(2019, 11).daysInMonth(), 30);
    expect(DateCalc(2019, 12).daysInMonth(), 31);
    expect(DateCalc(2020, 2).daysInMonth(), 29);
    expect(DateCalc.daysInMonthOf(year: 2020, month: 2), 29);
  });

  test('beginningOfDay returns beginning of the day', () {
    final t = DateCalc.now();
    expect(t.beginningOfDay() == DateTime(t.year, t.month, t.day), true);
    expect(t.beginningOfDay() == DateTime.now(), false);
  });

  test('endOfDay returns end of the day', () {
    final t = DateCalc.now();
    expect(t.endOfDay() == DateTime(t.year, t.month, t.day), false);
    expect(
        t.endOfDay().difference(t.beginningOfDay()) ==
            const Duration(hours: 23, minutes: 59, seconds: 59, milliseconds: 999),
        true);
  });

  test('beginningOfMonth returns beginning of the month', () {
    final t1 = DateCalc.now();
    final t2 = DateCalc(2020, 2, 20);
    expect(t1.beginningOfMonth() == DateTime(t1.year, t1.month), true);
    expect(t1.beginningOfMonth() == DateTime.now(), false);
    expect(t2.beginningOfMonth() == DateTime(2020, 2, 1), true);
  });

  test('endOfMonth returns end of the month', () {
    final t1 = DateCalc(2019, 2, 1);
    final t2 = DateCalc(2020, 2, 1);
    expect(t1.endOfMonth().month == 2, true);
    expect(t1.endOfMonth().day == 28, true);
    expect(t2.endOfMonth().month == 2, true);
    expect(t2.endOfMonth().day == 29, true);
    expect(
        t2.endOfMonth().difference(t2.beginningOfMonth()) ==
            const Duration(
                days: 28,
                hours: 23,
                minutes: 59,
                seconds: 59,
                milliseconds: 999),
        true);
  });

  test('utc constructs utc', () {
    final t = DateCalc.now();
    expect(DateCalc.utc(t.year, t.month, t.day).isUtc, true);
    expect(DateCalc(t.year, t.month, t.day).isUtc, false);
  });

  test('dup constructs dup', () {
    final t = DateCalc.now();
    expect(t.dup() == t, true);
    expect(t.dup(day: t.day + 1) == t.add(const Duration(days: 1)), true);
  });

  test('parse returns correct DateCalc instance', () {
    expect(DateCalc.parse('2019-01-02 03:45:06') is DateCalc, true);
    expect(
        DateCalc.parse('2019-01-02 03:45:06') ==
            DateTime.parse('2019-01-02 03:45:06'),
        true);
  });

  test('add years go well', () {
    expectDate(DateCalc(2010, 1, 1).addYear(1), 2011, 1, 1);
    expectDate(DateCalc(2019, 1, 1).addYear(1), 2020, 1, 1);
    expectDate(DateCalc(2020, 2, 29).addYear(1), 2021, 2, 28);
  });

  test('add months go well', () {
    expectDate(DateCalc(2019, 12, 1).addMonth(1), 2020, 1, 1);
    expectDate(DateCalc(2019, 1, 1).addMonth(1), 2019, 2, 1);
    expectDate(DateCalc(2019, 1, 31).addMonth(1), 2019, 2, 28);
    expectDate(DateCalc(2020, 1, 31).addMonth(1), 2020, 2, 29);
    expectDate(DateCalc(2019, 1, 31).addMonth(13), 2020, 2, 29);
  });

  test('add days go well', () {
    final t = DateCalc.now();
    expect(t.addDay(1) == t.add(const Duration(days: 1)), true);
    expect(t.addDay(31) == t.add(const Duration(days: 31)), true);
    expectDate(DateCalc(2019, 2, 28).addDay(365), 2020, 2, 28);
    expectDate(DateCalc(2019, 2, 28).addDay(366), 2020, 2, 29);
    expectDate(DateCalc(2019, 2, 28).addDay(367), 2020, 3, 1);
  });

  test('add hours go well', () {
    final t = DateCalc.now();
    expect(t.addHour(1) == t.add(const Duration(hours: 1)), true);
    expect(t.addHour(24) == t.add(const Duration(hours: 24)), true);
    expect(DateCalc(2019, 2, 28).addHour(1) == DateCalc(2019, 2, 28, 1), true);
  });

  test('add minutes go well', () {
    final t = DateCalc.now();
    expect(t.addMinute(1) == t.add(const Duration(minutes: 1)), true);
    expect(t.addMinute(60) == t.add(const Duration(minutes: 60)), true);
    expect(DateCalc(2019, 2, 28).addMinute(1) == DateCalc(2019, 2, 28, 0, 1),
        true);
  });

  test('add seconds go well', () {
    final t = DateCalc.now();
    expect(t.addSecond(1) == t.add(const Duration(seconds: 1)), true);
    expect(t.addSecond(60) == t.add(const Duration(seconds: 60)), true);
    expect(DateCalc(2019, 2, 28).addSecond(1) == DateCalc(2019, 2, 28, 0, 0, 1),
        true);
  });

  test('add milliseconds go well', () {
    final t = DateCalc.now();
    expect(t.addMillisecond(1) == t.add(const Duration(milliseconds: 1)), true);
    expect(t.addMillisecond(60) == t.add(const Duration(milliseconds: 60)), true);
    expect(
        DateCalc(2019, 2, 28).addMillisecond(1) ==
            DateCalc(2019, 2, 28, 0, 0, 0, 1),
        true);
  });

  test('add microseconds go well', () {
    final t = DateCalc.now();
    expect(t.addMicrosecond(1) == t.add(const Duration(microseconds: 1)), true);
    expect(t.addMicrosecond(60) == t.add(const Duration(microseconds: 60)), true);
    expect(
        DateCalc(2019, 2, 28).addMicrosecond(1) ==
            DateCalc(2019, 2, 28, 0, 0, 0, 0, 1),
        true);
  });

  test('subtract years go well', () {
    expectDate(DateCalc(2011, 1, 1).subtractYear(1), 2010, 1, 1);
    expectDate(DateCalc(2020, 1, 1).subtractYear(1), 2019, 1, 1);
    expectDate(DateCalc(2020, 2, 29).subtractYear(1), 2019, 2, 28);
  });

  test('subtract months go well', () {
    expectDate(DateCalc(2020, 1, 31).subtractMonth(1), 2019, 12, 31);
    expectDate(DateCalc(2020, 5, 31).subtractMonth(3), 2020, 2, 29);
    expectDate(DateCalc(2020, 2, 29).subtractMonth(12), 2019, 2, 28);
  });

  test('subtract days go well', () {
    final t = DateCalc.now();
    expect(t.subtractDay(1) == t.subtract(const Duration(days: 1)), true);
    expect(t.subtractDay(31) == t.subtract(const Duration(days: 31)), true);
    expectDate(DateCalc(2020, 3, 1).subtractDay(367), 2019, 2, 28);
    expectDate(DateCalc(2020, 2, 29).subtractDay(366), 2019, 2, 28);
    expectDate(DateCalc(2020, 2, 28).subtractDay(365), 2019, 2, 28);
  });

  test('subtract hours go well', () {
    final t = DateCalc.now();
    expect(t.subtractHour(1) == t.subtract(const Duration(hours: 1)), true);
    expect(t.subtractHour(24) == t.subtract(const Duration(hours: 24)), true);
    expect(DateCalc(2020, 3, 1).subtractHour(1) == DateCalc(2020, 2, 29, 23),
        true);
  });

  test('subtract minutes go well', () {
    final t = DateCalc.now();
    expect(t.subtractMinute(1) == t.subtract(const Duration(minutes: 1)), true);
    expect(t.subtractMinute(60) == t.subtract(const Duration(minutes: 60)), true);
    expect(
        DateCalc(2020, 3, 1).subtractMinute(1) == DateCalc(2020, 2, 29, 23, 59),
        true);
  });

  test('subtract seconds go well', () {
    final t = DateCalc.now();
    expect(t.subtractSecond(1) == t.subtract(const Duration(seconds: 1)), true);
    expect(t.subtractSecond(60) == t.subtract(const Duration(seconds: 60)), true);
    expect(
        DateCalc(2020, 3, 1).subtractSecond(1) ==
            DateCalc(2020, 2, 29, 23, 59, 59),
        true);
  });

  test('subtract milliseconds go well', () {
    final t = DateCalc.now();
    expect(t.subtractMillisecond(1) == t.subtract(const Duration(milliseconds: 1)),
        true);
    expect(
        t.subtractMillisecond(1000) == t.subtract(const Duration(milliseconds: 1000)),
        true);
    expect(
        DateCalc(2020, 3, 1).subtractMillisecond(1) ==
            DateCalc(2020, 2, 29, 23, 59, 59, 999),
        true);
  });

  test('subtract microseconds go well', () {
    final t = DateCalc.now();
    expect(t.subtractMicrosecond(1) == t.subtract(const Duration(microseconds: 1)),
        true);
    expect(
        t.subtractMicrosecond(1000) == t.subtract(const Duration(microseconds: 1000)),
        true);
    expect(
        DateCalc(2020, 3, 1).subtractMicrosecond(1) ==
            DateCalc(2020, 2, 29, 23, 59, 59, 999, 999),
        true);
  });

  test('differenceValue returns correct year value', () {
    expect(
        DateCalc(2020, 1, 1)
            .differenceValue(date: DateTime(2020, 1, 2), type: DateType.year),
        0);
    expect(
        DateCalc(2020, 1, 1)
            .differenceValue(date: DateTime(2021, 1, 2), type: DateType.year),
        1);
    expect(
        DateCalc(2020, 1, 3)
            .differenceValue(date: DateTime(2021, 1, 2), type: DateType.year),
        0);
    expect(
        DateCalc(2020, 4, 5)
            .differenceValue(date: DateTime(2019, 1, 2), type: DateType.year),
        1);
  });

  test('differenceValue returns correct month value', () {
    expect(
        DateCalc(2020, 1, 1)
            .differenceValue(date: DateTime(2020, 1, 2), type: DateType.month),
        0);
    expect(
        DateCalc(2020, 1, 1)
            .differenceValue(date: DateTime(2021, 1, 2), type: DateType.month),
        12);
    expect(
        DateCalc(2020, 1, 3)
            .differenceValue(date: DateTime(2021, 1, 2), type: DateType.month),
        11);
    expect(
        DateCalc(2020, 4, 5)
            .differenceValue(date: DateTime(2019, 1, 2), type: DateType.month),
        15);
  });

  test('differenceValue returns correct day value', () {
    expect(
        DateCalc(2020, 1, 1, 13, 45).differenceValue(
            date: DateTime(2020, 1, 2, 12, 50), type: DateType.day),
        0);
    expect(
        DateCalc(2020, 1, 1, 12, 45).differenceValue(
            date: DateTime(2021, 1, 2, 12, 50), type: DateType.day),
        367);
    expect(
        DateCalc(2020, 2, 3, 12, 45).differenceValue(
            date: DateTime(2020, 3, 4, 12, 50), type: DateType.day),
        30);
    expect(
        DateCalc(2020, 1, 1)
            .differenceValue(date: DateTime(2020, 2, 1), type: DateType.day),
        31);
    expect(
        DateCalc(2020, 2, 1)
            .differenceValue(date: DateTime(2020, 1, 1), type: DateType.day),
        31);
  });

  test('differenceValue returns correct other values', () {
    expect(
        DateCalc(2020, 1, 1, 13, 45).differenceValue(
            date: DateTime(2020, 1, 1, 12, 50), type: DateType.hour),
        0);
    expect(
        DateCalc(2020, 1, 1, 13, 45).differenceValue(
            date: DateTime(2020, 1, 1, 12, 45), type: DateType.hour),
        1);
    expect(
        DateCalc(2020, 1, 1, 12, 45, 5).differenceValue(
            date: DateTime(2020, 1, 1, 12, 44, 50), type: DateType.minute),
        0);
    expect(
        DateCalc(2020, 1, 1, 12, 45, 5).differenceValue(
            date: DateTime(2020, 1, 1, 12, 44, 5), type: DateType.minute),
        1);
    expect(
        DateCalc(2020, 1, 1, 12, 45, 5, 30).differenceValue(
            date: DateTime(2020, 1, 1, 12, 45, 4, 50), type: DateType.second),
        0);
    expect(
        DateCalc(2020, 1, 1, 12, 45, 5, 50).differenceValue(
            date: DateTime(2020, 1, 1, 12, 45, 4, 50), type: DateType.second),
        1);
  });

  test('toDate returns DateTime instance', () {
    expect(DateCalc.now().toDate() is DateTime, true);
    expect(DateCalc.now().toDate() is DateCalc, false);
  });

  test('toDate returns DateTime instance with correct Utc', () {
    expect(DateCalc.utc(2000).toDate().isUtc, true);
    expect(DateCalc.now().toDate().isUtc, false);
  });

  test('toFormattedString returns formatted string', () {
    final t = DateCalc(2019, 1, 2, 3, 45);
    expect(t.toFormattedString(), 'January 2, 2019 3:45:00 AM');
    expect(t.toFormattedString('yMd'), '1/2/2019');
    initializeDateFormatting('ja_JP');
    expect(t.toFormattedString('MMMM', 'ja_JP'), '1月');
  });
}

void expectDate(DateTime date, int y, [int m = 1, int d = 1]) {
  expect(date, DateTime(y, m, d));
}


lib/date_calc.dart
library date_calc;

import 'package:intl/intl.dart';

enum DateType { year, month, day, hour, minute, second }

class UnexceptedMonthException implements Exception {
  String cause;
  UnexceptedMonthException(this.cause);
}

class UnexpectedDateTypeException implements Exception {
    String cause;
  UnexpectedDateTypeException(this.cause);
}

class DateCalc extends DateTime {
  static final _durationOneMs = const Duration(milliseconds: 1);
  /// Days count in a month.
  static final _daysInMonth = {
    1: 31,
    2: 28,
    3: 31,
    4: 30,
    5: 31,
    6: 30,
    7: 31,
    8: 31,
    9: 30,
    10: 31,
    11: 30,
    12: 31,
    'leap': 29,
  };

  /// Constructs a [DateCalc] instance specified in the local time zone.
  DateCalc(
    int year, [
    int month = 1,
    int day = 1,
    int hour = 0,
    int minute = 0,
    int second = 0,
    int millisecond = 0,
    int microsecond = 0,
  ]) : super(
          year,
          month,
          day,
          hour,
          minute,
          second,
          millisecond,
          microsecond,
        );

  /// Constructs a [DateCalc] instance specified in the UTC time zone.
  DateCalc.utc(
    int year, [
    int month = 1,
    int day = 1,
    int hour = 0,
    int minute = 0,
    int second = 0,
    int millisecond = 0,
    int microsecond = 0,
  ]) : super.utc(
          year,
          month,
          day,
          hour,
          minute,
          second,
          millisecond,
          microsecond,
        );

  /// Constructs a [DateCalc] instance from a [DateTime] instance.
  DateCalc.fromDateTime(DateTime date)
      : super(
          date.year,
          date.month,
          date.day,
          date.hour,
          date.minute,
          date.second,
          date.millisecond,
          date.microsecond,
        );

  /// Constructs a [DateCalc] instance with current date and time in the
  /// local time zone.
  DateCalc.now() : super.now();

  /// Constructs a [DateCalc] instance duplicated.
  /// When pass some arguments, the [DateCalc] instance reflects it.
  /// DateCalc(2020, 01, 01).dup() => 2020-01-01 00:00:00.000
  /// DateCalc(2020, 01, 01).dup(day: 2, hour: 1) => 2020-01-02 01:00:00.000
  DateCalc dup({
    int? year,
    int? month,
    int? day,
    int? hour,
    int? minute,
    int? second,
    int? millisecond,
    int? microsecond,
  }) {
    return DateCalc(
      year ?? this.year,
      month ?? this.month,
      day ?? this.day,
      hour ?? this.hour,
      minute ?? this.minute,
      second ?? this.second,
      millisecond ?? this.millisecond,
      microsecond ?? this.microsecond,
    );
  }

  /// Returns a [DateCalc] instance based on [formattedString].
  /// DateCalc.parse('2019-01-02 03:45:06')
  /// => 2019-01-02 03:45:06.000
  // ignore: prefer_constructors_over_static_methods
  static DateCalc parse(String formattedString) {
    return DateCalc.fromDateTime(DateTime.parse(formattedString));
  }

  /// Returns true if given year is a leap year.
  /// DateCalc.isLeapYearFor(2020)
  /// => true
  static bool isLeapYearFor(int year) =>
      (year % 4 == 0) && ((year % 100 != 0) || (year % 400 == 0));

  /// Returns days count of given year and month.
  /// DateCalc.daysInMonthOf(year: 2020, month: 2)
  /// => 29
  static int daysInMonthOf({required int year, required int month}) {
          final daysInMonth = (month == DateTime.february && isLeapYearFor(year))
          ? _daysInMonth['leap']
          : _daysInMonth[month];
          if (daysInMonth != null) {
            return daysInMonth;
          }
          throw new UnexceptedMonthException('Unexpected month $month');
  }
      

  /// Returns a [DateCalc] instance with the beginning time of the day.
  /// DateCalc(2020, 2, 5).beginningOfDay()
  /// => 2020-02-05 00:00:00.000
  DateCalc beginningOfDay() => DateCalc(year, month, day);

  /// Returns a [DateCalc] instance with the end time of the day.
  /// End time means '23:59:59.999', not '23:59:59.999.999'.
  /// DateCalc(2020, 2, 5).endOfDay()
  /// => 2020-02-05 23:59:59.999
  DateCalc endOfDay() => DateCalc.fromDateTime(
      DateTime(year, month, day + 1).subtract(_durationOneMs));

  /// Returns a [DateCalc] instance with the beginning time of the month.
  /// DateCalc(2020, 2, 5).beginningOfMonth()
  /// => 2020-02-01 00:00:00.000
  DateCalc beginningOfMonth() => DateCalc(year, month);

  /// Returns a [DateCalc] instance with the end day and time of the month.
  /// End time means '23:59:59.999', not '23:59:59.999.999'.
  /// DateCalc(2020, 2, 5).endOfMonth()
  /// => 2020-02-29 23:59:59.999
  DateCalc endOfMonth() => DateCalc.fromDateTime(
      DateTime(year, month + 1).subtract(_durationOneMs));

  /// Returns true if self and given [date] are the same day.
  /// DateCalc(2020, 2, 5).isSameDay(DateTime(2020, 2, 5))
  /// => true
  bool isSameDay(DateTime date) =>
      year == date.year && month == date.month && day == date.day;

  /// Returns true if self and given [date] are the different day.
  /// DateCalc(2020, 2, 5).isNotSameDay(DateTime(2020, 2, 5))
  /// => false
  bool isNotSameDay(DateTime date) => !isSameDay(date);

  /// Returns true if self is today.
  /// DateTime.now() => 2020-02-05 00:00:00.000
  /// DateCalc(2020, 2, 5).isToday()
  /// => true
  bool isToday() => isSameDay(DateTime.now());

  /// Returns true if self is a leap year.
  /// DateCalc(2020, 2, 5).isLeapYearFor()
  /// => true
  bool isLeapYear() => isLeapYearFor(year);

  /// Returns days count of self.
  /// DateCalc(2020, 2, 5).daysInMonthOf()
  /// => 29
  int daysInMonth() => daysInMonthOf(year: year, month: month);

  /// Returns a [DateCalc] instance added years.
  /// If there is no corresponding day, returns the end day of the month insted.
  /// DateCalc(2020, 2, 29).addYear(1)
  /// => 2021-02-28 00:00:00.000
  DateCalc addYear(int other) {
    final result = dup(year: year + other);
    return result.month == month
        ? result
        : result.dup(
            month: month,
            day: daysInMonthOf(year: result.year, month: result.month - 1),
          );
  }

  /// Returns a [DateCalc] instance added months.
  /// If there is no corresponding day, returns the end day of the month insted.
  /// DateCalc(2020, 1, 31).addMonth(1)
  /// => 2020-02-29 00:00:00.000
  DateCalc addMonth(int other) {
    final result = dup(month: month + other);
    return result.day == day
        ? result
        : result.dup(
            month: result.month - 1,
            day: daysInMonthOf(year: result.year, month: result.month - 1),
          );
  }

  /// Returns a [DateCalc] instance added days.
  DateCalc addDay(int other) => dup(day: day + other);

  /// Returns a [DateCalc] instance added hours.
  DateCalc addHour(int other) => dup(hour: hour + other);

  /// Returns a [DateCalc] instance added minutes.
  DateCalc addMinute(int other) => dup(minute: minute + other);

  /// Returns a [DateCalc] instance added seconds.
  DateCalc addSecond(int other) => dup(second: second + other);

  /// Returns a [DateCalc] instance added milliseconds.
  DateCalc addMillisecond(int other) => dup(millisecond: millisecond + other);

  /// Returns a [DateCalc] instance added microseconds.
  DateCalc addMicrosecond(int other) => dup(microsecond: microsecond + other);

  /// Returns a [DateCalc] instance subtracted years.
  /// If there is no corresponding day, returns the end day of month insted.
  /// DateCalc(2020, 2, 29).subtractYear(1)
  /// => 2019-02-28 00:00:00.000
  DateCalc subtractYear(int other) {
    final result = dup(year: year - other);
    return result.month == month
        ? result
        : result.dup(
            month: month,
            day: daysInMonthOf(year: result.year, month: result.month - 1),
          );
  }

  /// Returns a [DateCalc] instance subtracted months.
  /// If there is no corresponding day, returns the end day of month insted.
  /// DateCalc(2020, 3, 31).subtractMonth(1)
  /// => 2020-02-29 00:00:00.000
  DateCalc subtractMonth(int other) {
    final result = dup(month: month - other);
    return result.day == day
        ? result
        : result.dup(
            month: result.month - 1,
            day: daysInMonthOf(year: result.year, month: result.month - 1),
          );
  }

  /// Returns a [DateCalc] instance subtracted days.
  DateCalc subtractDay(int other) => dup(day: day - other);

  /// Returns a [DateCalc] instance subtracted hours.
  DateCalc subtractHour(int other) => dup(hour: hour - other);

  /// Returns a [DateCalc] instance subtracted minutes.
  DateCalc subtractMinute(int other) => dup(minute: minute - other);

  /// Returns a [DateCalc] instance subtracted seconds.
  DateCalc subtractSecond(int other) => dup(second: second - other);

  /// Returns a [DateCalc] instance subtracted milliseconds.
  DateCalc subtractMillisecond(int other) =>
      dup(millisecond: millisecond - other);

  /// Returns a [DateCalc] instance subtracted microseconds.
  DateCalc subtractMicrosecond(int other) =>
      dup(microsecond: microsecond - other);

  /// Returns int of given type.
  /// When no given date, calculate self and DateTime.now().
  /// DateCalc(2020, 1, 5).differenceValue(date: DateTime(2020, 3, 3), type: DateType.month)
  /// => 1
  /// DateTime.now() => 2020, 2, 3
  /// DateCalc(2020, 1, 2).differenceValue(type: DateType.month)
  /// => 1
  int differenceValue({DateTime? date, required DateType type}) {
    final other = date == null ? DateCalc.now() : DateCalc.fromDateTime(date);
    final e = isBefore(other) ? this : other;
    final l = isBefore(other) ? other : this;
    int result;

    switch (type) {
      case DateType.year:
        result = (l.year - e.year) + (e.dup(year: l.year).isAfter(l) ? -1 : 0);
        break;
      case DateType.month:
        result = (l.year - e.year) * 12 +
            (l.month - e.month) +
            (e.dup(year: l.year, month: l.month).isAfter(l) ? -1 : 0);
        break;
      case DateType.day:
        result = l.difference(e).inDays;
        break;
      case DateType.hour:
        result = l.difference(e).inHours;
        break;
      case DateType.minute:
        result = l.difference(e).inMinutes;
        break;
      case DateType.second:
        result = l.difference(e).inSeconds;
        break;
      default:
        throw new UnexpectedDateTypeException('Unexpected DateType of $type');
    }
    return result;
  }

  /// Returns a [DateTime] instance.
  DateTime toDate() {
    return isUtc
        ? DateTime.utc(
            year, month, day, hour, minute, second, millisecond, microsecond)
        : DateTime(
            year, month, day, hour, minute, second, millisecond, microsecond);
  }

  /// Returns fomatted string by [DateFormat].
  /// This method provides simply parsing.
  /// When you need some extra methods like parseLoose or add_m,
  /// you can use [DateFormat] itself.
  ///
  /// DateCalc(2019, 2, 1).toFormattedString()
  /// => February 1, 2019 12:00:00 AM
  /// DateCalc(2019, 2, 1).toFormattedString('yMd')
  /// => 2/1/2019
  /// DateCalc(2019, 2, 1).toFormattedString('yMd', 'ja_JP')
  /// => 2019/2/1
  ///
  /// You can check usable locale in dateTimePatternMap
  /// in intl/date_time_patterns.dart.
  /// Call [initializeDateFormatting(locale)], then do [toFormattedString].
  String toFormattedString([String? format, String? locale]) {
    return DateFormat(format, locale).format(this);
  }
}
